daft/table/table_io.py

```diff
@@ -638,41 +588,88 @@ def file_visitor(written_file, protocol=protocol):
     target_file_size = 512 * 1024 * 1024
     TARGET_ROW_GROUP_SIZE = 128 * 1024 * 1024
 
-    arrow_table = mp.to_arrow()
+    format = pads.ParquetFileFormat()
+
+    opts = format.make_write_options(compression="zstd", use_compliant_nested_type=False)
 
     file_schema = schema_to_pyarrow(schema)
 
-    # This ensures that we populate field_id for iceberg as well as fill in null values where needed
-    # This might break for nested fields with large_strings
-    # we should test that behavior
-    arrow_table = coerce_pyarrow_table_to_schema(arrow_table, file_schema)
+    data_files = []
 
-    size_bytes = arrow_table.nbytes
+    for arrow_table, path, partition in micropartition_to_arrow_tables(mp, resolved_path, schema, partition_spec):
 
-    target_num_files = max(math.ceil(size_bytes / target_file_size / inflation_factor), 1)
-    num_rows = len(arrow_table)
+        def file_visitor(written_file, protocol=protocol):
```

Comment:
Maybe a good idea to pull this out as a callable `class` to avoid instantiation of a function on every iteration of the loop. Looks to me like it's just parametrized by `partition`?

Reply:
Would a higher order function work as well? One that takes in a `partition` and returns a `file_visitor`

Reply:
Actually it feels cleaner to just do a class I think

Reply:
Class would be preferred here so we can explicitly keep state I would think